File: examples/mini_lang/ast_nodes.py
"""AST node value types. Deliberately a small closed set -> Rust enum + match."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Number:
    value: float


@dataclass
class Var:
    name: str


@dataclass
class UnaryOp:
    op: str  # "-"
    operand: "Expr"


@dataclass
class BinOp:
    op: str  # "+" "-" "*" "/"
    left: "Expr"
    right: "Expr"


# An expression is one of the node types above.
Expr = "Number | Var | UnaryOp | BinOp"


@dataclass
class LetStmt:
    name: str
    expr: "Expr"


@dataclass
class ExprStmt:
    expr: "Expr"


# A statement is a binding or a bare expression.
Stmt = "LetStmt | ExprStmt"

File: examples/mini_lang_rust/src/ast_nodes.rs
//! AST node types. A small closed set -> idiomatic Rust enum + match.

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Number(f64),
    Var(String),
    Unary(Box<Expr>),                   // unary minus
    Binary(char, Box<Expr>, Box<Expr>), // op in {'+','-','*','/'}
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    Let(String, Expr),
    Expr(Expr),
}
